### PR TITLE
added implementation of NSGA-II

### PR DESCRIPTION
**Pull request by TennisGazelle**
Brief Overview:
---------------
image

Detailed Overview:
-------------------
 - Added sorting in population for each obj. function
 - Fast ND Sort of `ParedoFront`
   - `ParedoFront` which is a child of `population` is sorted
 - Crowding Distance assignment of `ParedoFront`
 - Crowding Comparison Operator of `Individual`s
 - `ParedoDominates` comparator for `Individual`s
 - `GA` holds NSGA implementation in `NSGAStep()` and `NSGARun()`

### CODE REVIEW

File: twitter-subset-selection/include/Population.h [hunk 1]
```diff
@@ -29,4 +30,14 @@ class Population : public vector<Individual> {
     double sumFitness;
 };
 
+class ParedoFront : public Population {
```
Comment (harpreets652) [inline twitter-subset-selection/include/Population.h]: nitpicking: it's Pareto

File: twitter-subset-selection/src/Individual.cpp [hunk 2]
```diff
@@ -61,4 +62,15 @@ void Individual::mutate() {
     (*this)[index] = !at((unsigned int)index);
 }
 
+bool Individual::paredoDominates(const Individual &opponent) const {
+    return (this->accuracy > opponent.accuracy) && (this->numFeaturesActive < opponent.numFeaturesActive);
```
Comment (harpreets652) [inline twitter-subset-selection/src/Individual.cpp]: This is not correct: both objective functions don't have to be better. Pareto Dominates: one objective is strictly greater and the other is greater than or equal to. The negated conditional might be easier to code: 
 !(this->accuracy < opponent.accuracy) && !(this->numFeaturesActive > opponent.numFeaturesActive)

Comment (TennisGazelle) [reply twitter-subset-selection/src/Individual.cpp]: oh shit this code is easier

Comment (TennisGazelle) [reply twitter-subset-selection/src/Individual.cpp]: wait, don't we need at least one strictly greater (or less) than?  The code you propose lets both these values be equal to one another.

Comment (harpreets652) [reply twitter-subset-selection/src/Individual.cpp]: True but I think the negation of that statement is enough to check for dominance. For example, consider the following: this = (95, 10), opponent = (92, 8); in this case this does not dominate opponent, right (this's accuracy is greater(which we want to maximize) but num of features is greater(which we want to minimize))? And just making sure that accuracy is not less and num of features is not greater yields the same result: this does not dominate opponent. 

Another example in which This does dominate Opponent: This = (95, 10), Opponent = (95, 12). Again checking the negation yields the same result: 
!(this.accuracy < opponent.accuracy) = true 
!(this.numFeatures > opponent.numFeatures) = true.

Can you think of a scenario the negation won't work?

Comment (TennisGazelle) [reply twitter-subset-selection/src/Individual.cpp]: You got me there.

**Pull request merged**